pkg/metadata/helper.go

```diff
@@ -0,0 +1,114 @@
+package metadata
+
+import (
+	"fmt"
+	"runtime"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/serializer"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	// run the host metadata collector every 1800 seconds (30 minutes)
+	hostMetadataCollectorInterval = 1800
+	// run the host metadata collector interval can be set through configuration within acceptable bounds
+	hostMetadataCollectorMinInterval = 300   // 5min minimum
+	hostMetadataCollectorMaxInterval = 14400 // 4h maximum
+	// run the agent checks metadata collector every 600 seconds (10 minutes)
+	agentChecksMetadataCollectorInterval = 600
+	// run the resources metadata collector every 300 seconds (5 minutes) by default, configurable
+	resourcesMetadataCollectorInterval = 300
+)
+
+type collector struct {
+	os          string
+	interval    time.Duration
+	min         time.Duration
+	max         time.Duration
+	ignoreError bool
+}
+
+var (
+	// default collectors by os
+	defaultCollectors = map[string]collector{
```

Comment:
these are the default collectors of the Agent specifically (they don't really apply to Dogstatsd). How about defining them in `cmd/agent/`, and passing them to `SetupMetadataCollection`? I _think_ you could then make dogstatsd use `SetupMetadataCollection` as well, or am I missing something?

Reply:
The agent and dogstatsd share the same logic (same default interval, min, max ,...). I would like to avoid duplicating the data when we do want to keep the same behavior.

Plus I think this logic belongs in the `metadata` package and not outside. It's up to the `metadata` package to make we send the payloads in an healthy way.

And it's up to each binary to know which provider they want to start by default.

WDYT ?

Reply:
I agree with both of your points, but to me it looks like the code in `cmd/dogstatsd` is doing too much compared to the code in `cmd/agent`. To me:

* the only thing the code in `cmd/dogstatsd` _should_ be doing is define that it only wants the `host` provider by default
* the only thing the code in `cmd/agent` should be doing is define that it wants "all" the default providers; or alternatively that it wants `host`, `agent_checks` and `resources`